statistics/histogram.go

```diff
@@ -1459,3 +1459,286 @@ func (hg *Histogram) ExtractTopN(cms *CMSketch, topN *TopN, numCols int, numTopN
 	topN.Sort()
 	return nil
 }
+
+// bucket4Merging is only used for merging partition hists to global hist.
+type bucket4Merging struct {
+	lower  *types.Datum
+	upper  *types.Datum
+	repeat int64
+	ndv    int64
+	count  int64
+	// disjointNDV is used for merging bucket NDV, see mergeBucketNDV for more details.
+	disjointNDV int64
+}
+
+func newBucket4Meging() *bucket4Merging {
+	return &bucket4Merging{
+		lower:  new(types.Datum),
+		upper:  new(types.Datum),
+		repeat: 0,
+		ndv:    0,
+		count:  0,
+	}
+}
+
+// buildBucket4Merging builds bucket4Merging from Histogram
+// Notice: Count in Histogram.Buckets is prefix sum but in bucket4Merging is not.
+func (hg *Histogram) buildBucket4Merging(buckets *[]*bucket4Merging) {
+	for i := 0; i < hg.Len(); i++ {
+		b := newBucket4Meging()
+		hg.GetLower(i).Copy(b.lower)
+		hg.GetUpper(i).Copy(b.upper)
+		b.repeat = hg.Buckets[i].Repeat
+		b.ndv = hg.Buckets[i].NDV
+		b.count = hg.Buckets[i].Count
+		if i != 0 {
+			b.count -= hg.Buckets[i-1].Count
+		}
+		*buckets = append(*buckets, b)
+	}
+}
+
+func (b *bucket4Merging) Clone() bucket4Merging {
+	return bucket4Merging{
+		lower:  b.lower.Clone(),
+		upper:  b.upper.Clone(),
+		repeat: b.repeat,
+		ndv:    b.ndv,
+		count:  b.count,
+	}
+}
+
+// mergeBucketNDV merges bucket NDV from tow bucket `b` & `left`.
+// Before merging, you need to make sure that when using (upper, lower) as the comparison key, `b` is greater than `left`
+func (b *bucket4Merging) mergeBucketNDV(sc *stmtctx.StatementContext, left *bucket4Merging) error {
```

Comment:
How about changing this method to a statistics function like `mergeBucketNDV(sc, left, right) (ndv, err)`, which is more easy-to-use and test?